# chapter - 02 자료형

## 2.1 자료형과 문자열

프로그래밍에서는 프로그램이 처리할 수 있는 모든 것을 자료(data)라고 부름.

### 자료형과 기본 자료형

파이썬 프로그램에서 수많은 자료를 다룸. 개발자들이 쉽게 사용할 수 있도록 기능과 역할에 따라서 자료를 구분. -> 자료형(data type)이라고 함.  
- 문자열(string) : 메일, 제목, 메시지 내용 등
- 숫자(number) : 물건의 가격, 학생의 성적 등
- 불(boolean) : 친구의 로그인 상태 등

#### 자료를 알아야 하는 이유

자료를 자료형에 맞게 모으고, 처리 과정을 거쳐 차근차근 더 큰 자료형을 만들어 나가다 보면 거대한 프로글매이 만들어짐.

#### 자료형 확인하기

자료형이란 자료의 형식을 말함. 파이썬에서 자료의 형식을 확인할 때는 type() 함수를 이용함.

In [1]:
print(type("안녕하세요"))

print(type(273))

<class 'str'>
<class 'int'>


string -> str 문자열  
integer -> int 정수

### 문자열 만들기

프로그래밍 언어에서는 글자들이 이처럼 나열된 것을 '문자열'이라고 부름. 문자열은 영어로 string(스트링)이라고 부름

> "Hello" 'String' '안녕하세요' "Hello Python Programming"

#### 큰 따옴표로 문자열 만들기

In [2]:
print("안녕하세요")

안녕하세요


#### 작은 따옴표로 문자열 만들기

In [3]:
print('안녕하세요')

안녕하세요


#### 문자열 내부에 따옴표 넣기

"안녕하세요"라고 말했습니다.

In [4]:
print(""안녕하세요"라고 말했습니다.")

SyntaxError: invalid syntax (2168074507.py, line 1)

파이썬 프로그래밍 언어는 자료(문자열)와 자료(문자열)를 단순하게 나열할 수 없음. 구문 오류 (Syntax Error) 발생 -> 작성한 코드에 문제가 발생해 실행조차 되지 않는다는 의미.

In [5]:
print('"안녕하세요"라고 말했습니다')

"안녕하세요"라고 말했습니다


In [6]:
print("'배가 고픕니다'라고 생각했습니다")

'배가 고픕니다'라고 생각했습니다


#### 이스케이프 문자를 사용해 문자열 만들기

이스케이프 문자(escape character)는 역슬래시(\) 기호와 함께 조합해서 사용하는 특수한 문자를 의미.

- \" : 큰 따옴표를 의미
- \' : 작은 따옴표를 의미

In [7]:
print("\"안녕하세요\"라고 말했습니다")
print('\'배가 고픕니다\'라고 생각했습니다')

"안녕하세요"라고 말했습니다
'배가 고픕니다'라고 생각했습니다


- \n : 줄바꿈을 의미함.
- \t : 탭을 의미.

In [8]:
print("안녕하세요\n안녕하세요")

print("안녕하세요\t안녕하세요")

안녕하세요
안녕하세요
안녕하세요	안녕하세요


##### 직접해보는 손코딩 (이스케이프 문자(\t)로 탭 사용하기)

In [9]:
print("이름\t나이\t지역")
print("윤인성\t25\t강서구")
print("윤아린\t24\t강서구")
print("구름\t3\t강서구")

이름	나이	지역
윤인성	25	강서구
윤아린	24	강서구
구름	3	강서구


- \\\ : 역슬래시(\)를 의미.

In [11]:
print("\\ \\ \\ \\")

\ \ \ \


#### 여러 줄 문자열 만들기

In [13]:
print("동해물과 백두산이 마르고 닳도록\n하느님이 보우하사 우리나라 만세\n무궁화 삼천리 화려강산 대한사람\n대한으로 길이 보전하세")

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세


파이썬은 여러 줄 문자열이라는 기능을 지원합니다. 여러 줄 문자열은 큰따옴표 또는 작은 따옴표를 세번 반복한 기호를 사용.

In [15]:
print("""동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세""")

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세


#### 줄바꿈 없이 문자열 만들기

In [17]:
print("""
동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세
""")


동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세



위아래로 의도하지 않은 줄바꿈이 들어가버림.  
줄바꿈을 하지 않고 코드를 입력하고 싶다면 다음과 같이 \ 기호를 사용.

In [18]:
print("""\
동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세\
""")

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산 대한사람
대한으로 길이 보전하세


### 문자열 연산자

'숫자'라는 자료에는 더하기, 빼기, 곱하기, 나누기라는 연산자를 적용.  
합집합, 교집합, 차집합이라는 연산자는 적용할 수 없음.

숫자에는 숫자에만 적용할 수 있는 연산자가 있고, 문자열에는 문자열에만 적용할 수 있는 연산자가 있음.

#### 문자열 연결 연산자 : +

"문자열" + "문자열"  
문자열 연결 연산자는 두 문자열을 연결해서 새로운 문자열을 만들어 내는 것.

In [19]:
print("안녕" + "하세요")

print("안녕하세요" + "!")

안녕하세요
안녕하세요!


In [20]:
print("안녕하세요" + 1)

TypeError: can only concatenate str (not "int") to str

문자열은 무조건 문자열끼리 + 기호를 사용해서 연결.  
숫자라 하더라도 문자열과 함께 + 연산하려면 큰따옴표를 붙여 문자열로 인식.   
숫자를 더할 때는 반드시 숫자와 숫자 사이에 + 기호를 사용해서 연산해야 함.

#### 문자열 반복 연산자 : *

In [22]:
print("안녕하세요" * 3)

안녕하세요안녕하세요안녕하세요


In [23]:
print(3 * "안녕하세요")

안녕하세요안녕하세요안녕하세요


#### 문자 선택 연산자(인덱싱):[]

문자 선택 연산자는 문자열 내부의 문자 하나를 선택하는 연산자. 대괄호[] 안에는 선택할 문자의 위치를 지정. 이 숫자를 인덱스(Index)라고 부름.  
  
숫자를 0부터 세는 제로 인덱스(zero index), 다른 하나는 숫자를 1부터 세는 원 인덱스(one index)로 구분. 파이썬은 제로 인덱스 사용.

##### 직접 해보는 손코딩 (문자 선택 연산자의 결과 출력하기)

In [25]:
print("문자 선택 연산자에 대해 알아볼까요?")
print("안녕하세요"[0])
print("안녕하세요"[1])
print("안녕하세요"[2])
print("안녕하세요"[3])
print("안녕하세요"[4])

문자 선택 연산자에 대해 알아볼까요?
안
녕
하
세
요


대괄호 안의 숫자를 음수로 입력하면 뒤에서부터 선택할 수 있음.

##### 직접해보는 손코딩 (뒤에서 부터 선택하기)

In [26]:
print("문자를 뒤에서부터 선택해 볼까요?")
print("안녕하세요"[-1])
print("안녕하세요"[-2])
print("안녕하세요"[-3])
print("안녕하세요"[-4])
print("안녕하세요"[-5])

문자를 뒤에서부터 선택해 볼까요?
요
세
하
녕
안


##### 문자열 범위 선택 연산자 (슬라이싱): [:]

In [27]:
print("안녕하세요"[1:4])

녕하세


파이썬은 '마지막 숫자를 포함하지 않음'으로 적용.

In [28]:
print("안녕하세요"[0:2])
print("안녕하세요"[1:3])
print("안녕하세요"[2:4])

안녕
녕하
하세


뒤의 값을 생략할 때는 자동으로 가장 최대 위치(마지막 글자)까지, 앞의 값을 생략할 때는 가장 앞쪽의 위치(첫번째 글자)까지 지정.  
[1:]  
[:3]

In [29]:
print("안녕하세요"[1:])
print("안녕하세요"[:3])

녕하세요
안녕하


[]기호를 이용해 문자열의 특정위치에 있는 문자를 참조하는 것을 인덱싱(indexing)이라고 함. [:] 기호를 이용해 문자열의 일부를 추출하는 것을 슬라이싱(slicing)이라고 함.  

  
문자열 선택 연산자로 슬라이스를 하더라도 원본은 변하지 않는 것을 기억해야 함.

In [31]:
hello = "안녕하세요"
print(hello[0:2])

안녕


In [32]:
hello

'안녕하세요'

#### IndexError(index out of range) 예외

프로그래밍을 할 때 가장 많이 만나는 예외 중에 하나를 꼽으라면 바로 IndexError 예외.  
IndexError 예외는 리스트 / 문자열의 수를 넘는 요소/글자를 선택할 때 발생.

In [33]:
print("안녕하세요"[10])

IndexError: string index out of range

### 문자열의 길이 구하기

문자열의 길이를 구할 때는 len() 함수를 사용.  
'문자열에 들어있는 문자의 개수(=문자열의 길이)'를 세어 줍니다.

In [35]:
print(len("안녕하세요"))

5


## 2.2 숫자

소수점이 없는 숫자를 정수형이라 하고, 소수점이 있는 숫자를 실수형이라 함.  
  
주의할 점은 0은 소수점이 없는 숫자이고, 0.0은 소수점이 있는 숫자.  
소수점이 있느냐에 따라서 숫자의 자료형이 달라지기 때문에 주의해야합니다.  
'소수점이 움직이는 숫자'라는 의미로 부동 소수점

### 숫자의 종류

In [36]:
print(273)

print(52.273)

273
52.273


type()은 괄호 안에 입력한 자료의 유형을 알려주는 함수

In [37]:
print(type(52))
print(type(52.273))

<class 'int'>
<class 'float'>


int는 integer의 앞 세 글자를 떼어낸 것으로 정수를 의미.  
float는 floating point의 앞 단어에서 가져온 것으로 부동 소수점(실수)를 의미.

In [38]:
print(0)
print(type(0))
print(0.0)
print(type(0.0))

0
<class 'int'>
0.0
<class 'float'>


In [39]:
print("5 + 7 = ", 5 + 7)
print("5 - 7 = ", 5 - 7)
print("5 * 7 = ", 5 * 7)
print("5 / 7 = ", 5 / 7)

5 + 7 =  12
5 - 7 =  -2
5 * 7 =  35
5 / 7 =  0.7142857142857143


#### 정수 나누기 연산자: //

숫자를 나누고 소수점 이하의 자릿수를 떼어 버린 후, 정수 부분만 남기는 정수 나누기 연산자.

In [40]:
print("3 / 2 = ", 3 / 2)
print("3 // 2 = ", 3 // 2)

3 / 2 =  1.5
3 // 2 =  1


// 연산자를 사용한 수식의 결과는 소수점 아래를 떼어 버린 값이 출력되는 것을 확인 할 수 있습니다.

#### 나머지 연산자 : %

나머지 연산자는 A를 B로 나누었을 때 남은 나머지를 구하는 연산자.

In [41]:
print("5 % 2 = ", 5 % 2)

5 % 2 =  1


#### 제곱 연산자 : **

파이썬에는 숫자를 제곱하는 **연산자가 있음.

In [44]:
print("2 ** 1 =", 2 ** 1)
print("2 ** 2 =", 2 ** 2)
print("2 ** 3 =", 2 ** 3)
print("2 ** 4 =", 2 ** 4)

2 ** 1 = 2
2 ** 2 = 4
2 ** 3 = 8
2 ** 4 = 16


### 연산자의 우선순위

In [47]:
print(2 + 2 - 2 * 2 / 2 * 2)
print(2 - 2 + 2 / 2 * 2 + 2)

0.0
4.0


In [48]:
(5 + 3) * 2

16

In [49]:
5 + (3 * 2)

11

#### TypeError 예외

서로 다른 자료를 연산하면 TypeError라는 예외가 발생.

In [50]:
string = "문자열"
number = 273
string + number

TypeError: can only concatenate str (not "int") to str

문자열은 + 연산자를 '문자열 연결 연산자'로 사용하려고 하고, 숫자는 + 연산자를 '덧셈 연산자'로 사용하려다 보니 충돌이 발생.

##### 문자열 연산자의 우선순위

문자열에 적용하는 연산자도 우선순위를 갖습니다.

문자열도 곱셈 연산자와 같은 * 기호가 우선한다는 것을 알 수 있음.

## 2.3 변수와 입력

변수는 값을 저장할 때 사용하는 식별자.

### 변수 만들기/사용하기

In [51]:
pi = 3.14159265
pi

3.14159265

##### 첫째, 변수를 선언하는 방법

##### 둘째, 변수에 값을 할당하는 방법

= 기호는 '같다'는 의미가 아니라 우변의 값을 좌변에 '넣겠다', '할당하겠다'를 의미.

##### 셋째, 변수를 참조하는 방법

변수에서 값을 꺼내는 것을 의미. 변수 안에 있는 값을 사용하는 것을 '변수 참조'

변수 = 값

In [52]:
pi = 3.14159265

- 변수에 저장된 값을 출력합니다.
> 변수
- 변수에 저장된 값으로 연산합니다.
> 변수 + 변수
- 변수에 저장된 값을 출력합니다.
> print(변수)

In [53]:
pi = 3.14159265
pi + 2

5.14159265

In [55]:
pi - 2

1.1415926500000002

In [57]:
pi * 2

6.2831853

In [59]:
pi / 2

1.570796325

In [60]:
pi % 2

1.1415926500000002

In [61]:
pi * pi

9.869604378534024

pi는 숫자 자료이기 때문에 숫자와 문자열을 연산하는 것은 당연히 안됩니다.
> pi + "문자열"

##### 직접 해보는 손코딩 (원의 둘레와 넓이 구하기)

In [62]:
# 변수 선언과 할당
pi = 3.14159265
r = 10

# 변수 참조
print("원주율 =", pi)
print("반지름 =", r)
print("원의 둘레 =", 2 * pi * r) # 원의 둘레
print("원의 넓이 =", pi * r * r) # 원의 넓이

원주율 = 3.14159265
반지름 = 10
원의 둘레 = 62.831853
원의 넓이 = 314.159265


##### 여기서 잠깐 (파이썬의 유연성)

자바, c의 경우에는 기본적으로 변수를 사용할 때 변수의 자료형에 대해 미리 선언해 주어야 함.

파이썬은 다른 프로그래밍 언어와는 다르게 변수에 자료형을 지정하지 않습니다. 이러한 유연성 때문에 변수에 어떠한 자료형이 들어 있는지 모르고 실수해서 실행 중에 TypeError를 발생할 확률이 높습니다.  
그러므로 하나의 변수에는 되도록 하나의 자료형을 넣어 활용하는 것이 좋습니다.

### 복합 대입 연산자

변수를 활용하면 기존의 연산자와 조합해서 사용할 수 있는 연산자가 있습니다. 이를 복합 대입 연산자라고 부릅니다.

복합 대입 연산자는 자료형에 적용하는 기본 연산자와 = 연산자를 함께 사용해 다음과 같이 구성하는 연산자.

> a += 10

In [63]:
number = 100
number += 10
number += 20
number += 30
print("number:", number)

number: 160


In [64]:
string = "안녕하세요"
string += "!"
string += "!"
print("string:", string)

string: 안녕하세요!!


### 사용자 입력 : input()

파이썬은 명령 프롬프트에서 사용자로부터 데이터를 입력받을 때 input() 함수를 사용.

#### input() 함수로 사용자 입력받기
input 함수 괄호 안에 입력한 내용을 프롬프트 문자열이라고 합니다.

In [69]:
input("인사말을 입력하세요> ")

'안녕하세요'

프로그램이 실행 도중 잠시 멈추는 것을 블록(block)이라고 함. input() 함수가 사용자에게 자료 입력을 요구하면서 코드 진행을 블록하고 있다고 할 수 있음.

In [71]:
string = input("인사말을 입력하세요> ")
print(string)

안녕하세요


input과 같이 함수의 결과로 나오는 값을 리턴값이라고 합ㄴ디ㅏ.

#### input() 함수의 입력 자료형

In [72]:
print(type(string))

<class 'str'>


In [73]:
number = input("숫자를 입력하세요> ")

print(number)

12345


In [74]:
print(type(number))

<class 'str'>


input()함수는 사용자가 무엇을 입력해도 결과는 무조건 문자열 자료형입니다. number에 대입한 12345도 'str'로 확인됩니다. True나 False와 같은 불(boolean)값도 모두 문자열로 나옴.

##### 직접 해보는 손코딩 (입력 자료형 확인하기)

In [77]:
# 입력을 받습니다
string = input("입력 > ")

# 출력합니다
print("자료:", string)
print("자료형", type(string))

자료: True
자료형 <class 'str'>


파이썬은 문자열은 문자열끼리 연산, 숫자는 숫자끼리 연산

##### 직접 해보는 손코딩 (입력받고 더하기)

In [78]:
# 입력을 받습니다
string = input("입력> ")

# 출력합니다
print("입력 + 100:", string + 100)


TypeError: can only concatenate str (not "int") to str

input()함수로 입력받은 자료는 모두 문자열로 저장되므로 "300" + 100이 되어 문자열과 숫자는 더할 수 없어 발생한 에러입니다.

### 문자열을 숫자로 바꾸기

input() 함수의 입력 자료형은 항상 문자열이기 때문에 입력받은 문장령르 숫자로 변환해야 숫자연산에 활용할 수 있음. -> 캐스트(cast)

- int() 함수 : 문자열을 int 자료형으로 변환합니다.
- float() 함수 : 문자열을 float 자료형으로 변환합니다.

##### 직접 해보는 손코딩 (int()함수 활용하기)

In [79]:
string_a = input("입력A> ")
int_a = int(string_a)

string_b = input("입력B> ")
int_b = int(string_b)

print("문자열 자료:", string_a + string_b)
print("숫자 자료:", int_a + int_b)

문자열 자료: 27352
숫자 자료: 325


##### 직접 해보는 손코딩 (int()함수와 float()함수 활용하기)

In [80]:
output_a = int("52")
output_b = float("52.273")

print(type(output_a), output_a)
print(type(output_b), output_b)

<class 'int'> 52
<class 'float'> 52.273


##### 직접 해보는 손코딩 (int()함수와 float()함수 조합하기)

In [81]:
input_a = float(input("첫 번째 숫자> "))
input_b = float(input("두 번째 숫자> "))

print("덧셈 결과:", input_a + input_b)
print("뺄셈 결과:", input_a - input_b)
print("곱셈 결과:", input_a * input_b)
print("나눗셈 결과:", input_a / input_b)

덧셈 결과: 325.0
뺄셈 결과: 221.0
곱셈 결과: 14196.0
나눗셈 결과: 5.25


#### ValueError 예외

자료형을 변환할 때 '변환할 수 없는 것'을 변환하려고 하면 ValueError 예외가 발생.

첫째, 숫자가 아닌 것을 숫자로 변환하려고 할 때
> int("안녕하세요")
> float("안녕하세요")

"안녕하세요"라는 문자열은 int() 함수로 변환할 수 없는 값이기 때문. int() 함수와 float() 함수는 매개변수로 변환할 수 없는 형태가 들어가면 항상 오류를 발생시킴.

둘째, 소수점이 있는 숫자 형식의 문자열을 int() 함수로 변환하려고 할 때
> int("52.273)

int는 정수인데, 부동 소수점이 있는 자료를 정수로 바꾸라 하면, 이 또한 에러가 발생.

##### 문제 해결

int() 함수는 부동 소수점 자료형을 정수로 변환하지 못하지만, float() 함수가 의미하는 실수는 정수도 포함하고 있기 때문에 정수, 실수 구분없이 사용할 수 있음.

### 숫자를 문자열로 바꾸기

str() 함수는 앞서 했던 int() 그리고 float() 함수와 비슷한 형태를 가집니다. 즉, 다른 자료형의 값을 str() 함수의 매개변수에 넣으면 문자열로 변환됩니다.
> str(다른 자료형)

##### 직접 해보는 손코딩 (str() 함수를 사용해 숫자를 문자열로 변환하기)

In [82]:
output_a = str(52)
output_b = str(52.273)
print(type(output_a), output_a)
print(type(output_b), output_b)

<class 'str'> 52
<class 'str'> 52.273


## 2.4 숫자와 문자열의 다양한 기능

문자열 뒤에 마침표(.)를 입력해 보면 자동 완성 기능으로 다양한 것들이 나옵니다.

### 문자열의 format() 함수

#### format()

format() 함수는 문자열이 가지고 있는 함수.  
중괄호{}를 포함한 문자열 뒤 마침표(.)를 찍고 format() 함수를 사용하는데, 중괄호의 개수와 format 함수 괄호 안 매개변수의 개수는 반드시 같아야 함.

> "{}".format(10)  
> "{} {}".format(10, 20)  
> "{} {} {} {} {}".format(101, 202, 303, 404, 505)

##### 직접 해보는 손코딩 (format()함수로 숫자를 문자열로 반환하기)

In [1]:
# format()함수로 숫자를 문자열로 변환하기
string_a = "{}".format(10)

# 출력하기
print(string_a)
print(type(string_a))

10
<class 'str'>


##### 직접 해보는 손코딩 (format()함수의 다양한 형태)

In [2]:
# format() 함수로 숫자를 문자열로 변환하기
format_a = "{}만 원".format(5000)
format_b = "파이썬 열공하여 첫 연봉 {}만 원 만들기".format(5000)
format_c = "{} {} {}".format(3000, 4000, 5000)
format_d = "{} {} {}".format(1, "문자열", True)

# 출력하기
print(format_a)
print(format_b)
print(format_c)
print(format_d)

5000만 원
파이썬 열공하여 첫 연봉 5000만 원 만들기
3000 4000 5000
1 문자열 True


##### IndexError 예외
{} 기호의 개수가 format()함수의 매개변수 개수보다 많으면 indexError 예외가 발생.  
매개변수가 {}보다 많은 경우로 {}개수 만큼 적용. 나머지 매개 변수는 버려짐.

In [4]:
"{} {}".format(1,2,3,4,5)

'1 2'

In [5]:
"{} {} {}".format(1,2)

IndexError: Replacement index 2 out of range for positional args tuple

### format() 함수의 다양한 기능

In [6]:
# 정수
output_a = "{:d}".format(52)

# 특정 칸에 처리하기
output_b = "{:5d}".format(52)   # 5칸
output_c = "{:10d}".format(52)  # 10칸

# 빈칸을 0으로 채우기
output_d = "{:05d}".format(52)  # 양수
output_e = "{:05d}".format(-52)  # 음수

print("# 기본")
print(output_a)

print("# 특정 칸에 출력하기")
print(output_b)
print(output_c)
print("# 빈칸을 0으로 채우기")
print(output_d)
print(output_e)

# 기본
52
# 특정 칸에 출력하기
   52
        52
# 빈칸을 0으로 채우기
00052
-0052


{:d}를 사용했을 때는 매개변수로 정수만 올 수 있습니다.  
{:5d}라고 입력하면 5칸을 잡고 뒤에서 부터 52라는 숫자를 채웁니다. {:10d}도 마찬가지로 10칸을 잡고 뒤에서 부터 52라는 숫자를 채움.  
{:05d}라고 지정하면 5칸을 잡고 뒤에서부터 52라는 숫자를 넣은 후, 앞의 빈 곳을 0으로 채움. 부호가 있을 때는 맨 앞자리를 부호로 채우고 나머지 빈곳을 0으로 채움.

##### 직접 해보는 손코딩 (기호 붙여 출력하기)

In [7]:
# 기호와 함께 출력하기
output_f = "{:+d}".format(52)   # 양수
output_g = "{:+d}".format(-52)   # 음수
output_h = "{: d}".format(52)   # 양수 : 기호 부분 공백
output_i = "{: d}".format(-52)   # 음수 : 기호 부분 공백

print("# 기호와 함께 출력하기")
print(output_f)
print(output_g)
print(output_h)
print(output_i)

# 기호와 함께 출력하기
+52
-52
 52
-52


##### 직접 해보는 손코딩 (조합해보기)

In [8]:
# 조합하기
output_h = "{:+5d}".format(52)  # 기호를 뒤로 밀기 : 양수
output_i = "{:+5d}".format(-52) # 기호를 뒤로 밀기 : 음수
output_j = "{:=+5d}".format(52) # 기호를 앞으로 밀기 : 양수
output_k = "{:=+5d}".format(-52)    # 기호를 앞으로 밀기 : 음수
output_l = "{:+05d}".format(52) # 0으로 채우기 : 양수
output_m = "{:+05d}".format(-52)    # 0으로 채우기 : 음수

print("# 조합하기")
print(output_h)
print(output_i)
print(output_j)
print(output_k)
print(output_l)
print(output_m)

# 조합하기
  +52
  -52
+  52
-  52
+0052
-0052


#### 부동 소수점 출력의 다양한 형태

##### 직접 해보는 손코딩 (float 자료형 기본)

In [9]:
output_a = "{:f}".format(52.273)
output_b = "{:15f}".format(52.273)  # 15칸 만들기
output_c = "{:+15f}".format(52.273) # 15칸에 부호 추가하기
output_d = "{:+015f}".format(52.273)    # 15칸 부호 추가하고 0으로 채우기

print(output_a)
print(output_b)
print(output_c)
print(output_d)

52.273000
      52.273000
     +52.273000
+0000052.273000


##### 직접 해보는 손코딩 (소수점 아래 자릿수 지정하기)

In [10]:
output_a = "{:15.3f}".format(52.273)
output_b = "{:15.2f}".format(52.273)
output_c = "{:15.1f}".format(52.273)

print(output_a)
print(output_b)
print(output_c)

         52.273
          52.27
           52.3


#### 의미 없는 소수점 제거하기

의미 없는 0을 제거한 후 출력하고 싶을 때는 {:g}를 사용함.
##### 직접해보는 손코딩 (의미 없는 소수점 제거하기)

In [11]:
output_a = 52.0
output_b = "{:g}".format(output_a)
print(output_a)
print(output_b)

52.0
52


### 대소문자 바꾸기: upper()와 lower()
upper()함수는 문자열의 알파벳을 대문자로, lower()함수는 문자열의 알파벳을 소문자로 만듬

In [12]:
a = "Hello Python Programming...!"
a.upper()

'HELLO PYTHON PROGRAMMING...!'

In [13]:
a.lower()

'hello python programming...!'

원본을 변화시키지 않는 함수를 비파괴적 함수. 이와 반대되는 함수를 파괴적 함수라 함.

### 문자열 양옆의 공백 제거하기: strip()
strip() 함수는 문자열 양옆의 공백을 제거  
문자열 양옆의 공백을 제거할 때는 strip()함수를 사용. 왼쪽 공백을 제거하는 lstrip() 함수와 오른쪽 공백을 제거하는 rstrip()함수도 있음.  
공백이란 '띄어쓰기','탭', '줄바꿈'을 모두 포함.  
- strip(): 문자열 양 옆의 공백을 제거
- lstrip(): 문자열 왼쪽 공백을 제거
- rstrip(): 문자열 오른쪽 공백을 제거

In [14]:
input_a = """
    안녕하세요
문자열의 함수를 알아봅니다
"""
print(input_a)


    안녕하세요
문자열의 함수를 알아봅니다



In [15]:
print(input_a.strip())

안녕하세요
문자열의 함수를 알아봅니다


코드를 실행한 결과를 보면 공백 제거 이전에는 양쪽에 줄바꿈과 띄어쓰기가 들어갔지만, 공백 제거 이후에는 이러한 것들이 모두 사라지는 것을 볼 수 있음.  
이러한 기능을 trim이라고 부름. 공백을 제거할 때는 strip 또는 trim을 활용한다고 기억.

### 문자열의 구성 파악하기: isOO()
문자열이 소문자로만 구성되어 있는지, 알파벳으로만 구성되어 있는지, 숫자로만 구성되어 있는지 등을 확인할 때는 is로 시작하는 이름의 함수를 사용.

- isalnum(): 문자열이 알파벳 또는 숫자로만 구성되어 있는지 확인.
- isalpha(): 문자열이 알파벳으로만 구성되어 있는지 확인.
- isidentifier(): 문자열이 식별자로 사용할 수 있는지 확인.
- isdecimal(): 문자열이 정수 형태인지 확인.
- isdigit(): 문자열이 숫자로 인식될 수 있는 것인지 확인.
- isspace(): 문자열이 공백으로만 구성되어 있는지 확인.
- islower(): 문자열이 소문자로만 구성되어 있는지 확인.
- isupper(): 문자열이 대문자로만 구성되어 있는지 확인.  
출력은 True(맞다) 또는 False(아니다)라고 나오는데, 이를 불(boolean)이라 부름.

In [16]:
print("TrainA10".isalnum())
print("10".isdigit())

True
True


### 문자열 찾기: find()와 rfind()
문자열 내부에 특정 문자가 어디에 위치하는지 확인 할 때 사용.
- find(): 왼쪽부터 찾아서 처음 등장하는 위치를 찾음
- rfind(): 오른쪽부터 찾아서 처음 등장하는 위치를 찾음

In [18]:
output_a = "안녕안녕하세요".find("안녕")
print(output_a)

0


In [19]:
output_b = "안녕안녕하세요".rfind("안녕")
print(output_b)

2


### 문자열과 in 연산자
문자열 내부에 어떤 문자열이 있는지 확인하려면 in 연산자를 사용.

In [20]:
print("안녕" in "안녕하세요")

True


In [21]:
print("잘자" in "안녕하세요")

False


### 문자열 자르기: split()
문자열을 특정한 문자로 자를 대는 split() 함수를 사용.

In [22]:
a = "10 20 30 40 50".split(" ")
print(a)

['10', '20', '30', '40', '50']
